### PR TITLE
MM-16258: Adds new API endpoint + (App & Client & Store) to retrieve …

### PR DESCRIPTION
**Pull request by mkraft**
…Users who would be removed from a list of hypothetical group IDs representing the synced groups.

#### Summary

For use by a [new system admin modal](https://projects.invisionapp.com/share/QHMWHPCCXYP#/screens/368106138), the client will provide:

* a `team_id` representing the current team being setup for group syncing
* a list of `group_ids` representing the groups that have been selected (but not yet saved) as the synced groups
* `page`
* `per_page`

The output is a payload containing:
* users that would be removed
* the groups they are members of
* their scheme roles
* the total count of results (because of pagination)

For example, this request:

```
GET /api/v4/teams/z19njpqwo3bnmfq966bp7cawiy/if_groups_then_users_removed?group_ids=eoezijg8zffgjmch8icy5bjd1e,ugaw6wjc3tfxpcr1eq5u5k8dhe&page=0&per_page=1
```

Could have this response:

```json
{
    "users": [
        {
            "id": "1gkib3xc9jrobqagxndgwf9hna",
            "create_at": 1560280685080,
            "update_at": 1560280685176,
            "delete_at": 0,
            "username": "lindy.test-simulator.amazonses.com",
            "auth_data": "lindy.test@simulator.amazonses.com",
            "auth_service": "ldap",
            "email": "lindy.test@simulator.amazonses.com",
            "email_verified": true,
            "nickname": "Lindy",
            "first_name": "Lindy",
            "last_name": "Test",
            "position": "",
            "roles": "system_user",
            "notify_props": {
                "channel": "true",
                "comments": "never",
                "desktop": "mention",
                "desktop_sound": "true",
                "email": "true",
                "first_name": "false",
                "mention_keys": "lindy.test-simulator.amazonses.com,@lindy.test-simulator.amazonses.com",
                "push": "mention",
                "push_status": "away"
            },
            "last_password_update": 1560280685080,
            "locale": "en",
            "timezone": {
                "automaticTimezone": "",
                "manualTimezone": "",
                "useAutomaticTimezone": "true"
            },
            "groups": [
                {
                    "id": "hw1pcpj7s3yijpfok6ssc8z7kh",
                    "name": "8b5iocstd78fzcsw8zkcwuzh7w",
                    "display_name": "engineering",
                    "description": "",
                    "source": "ldap",
                    "remote_id": "engineering",
                    "create_at": 1560280467586,
                    "update_at": 1560280467586,
                    "delete_at": 0,
                    "has_syncables": false
                }
            ],
            "scheme_guest": false,
            "scheme_user": true,
            "scheme_admin": false
        }
    ],
    "total_count": 2
}
``` 

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-16258

### CODE REVIEW

File: api4/team.go [hunk 1]
```diff
@@ -1082,3 +1086,42 @@ func updateTeamScheme(c *Context, w http.ResponseWriter, r *http.Request) {
...
 	ReturnStatusOK(w)
 }
+
+func ifGroupsThenUsersRemoved(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.RequireTeamId()
+	if c.Err != nil {
+		return
+	}
+
+	if len(c.Params.GroupIDs) < 26 {
```
Comment (reflog) [inline api4/team.go]: Nit 0/5: A constant would work better here than a magic number

Comment (mkraft) [reply api4/team.go]: I agree, and I'm actually surprised that one doesn't exist yet given that this is instance number 226 of this bad habit. Perhaps I'll submit a separate PR to fix all of them at once.

Comment (jespino) [reply api4/team.go]: Yes, I think we should have a constant for that, but I don't see adding it here. I like the idea of a whole PR replacing all the Id lengths checks with a single constant :)

File: api4/team.go [hunk 2]
```diff
@@ -58,6 +60,8 @@ func (api *API) InitTeam() {
 	api.BaseRoutes.Team.Handle("/invite/email", api.ApiSessionRequired(inviteUsersToTeam)).Methods("POST")
 	api.BaseRoutes.Teams.Handle("/invites/email", api.ApiSessionRequired(invalidateAllEmailInvites)).Methods("DELETE")
 	api.BaseRoutes.Teams.Handle("/invite/{invite_id:[A-Za-z0-9]+}", api.ApiHandler(getInviteInfo)).Methods("GET")
+
+	api.BaseRoutes.Teams.Handle("/{team_id:[A-Za-z0-9]+}/if_groups_then_users_removed", api.ApiSessionRequired(ifGroupsThenUsersRemoved)).Methods("GET")
```
Comment (jespino) [inline api4/team.go]: the url is a bit confusing to me, I don't think is a clear name. maybe `users_removed_on_changed_groups` `users_removed_if_groups_change`. I think is more clear an explicit.

File: app/group.go [hunk 3]
```diff
@@ -160,3 +162,58 @@ func (a *App) GetGroupsByTeam(teamId string, opts model.GroupSearchOpts) ([]*mod
 func (a *App) GetGroups(page, perPage int, opts model.GroupSearchOpts) ([]*model.Group, *model.AppError) {
 	return a.Srv.Store.Group().GetGroups(page, perPage, opts)
 }
+
+// IfGroupsThenTeamUsersRemoved returns the list of users who would be removed from the team if the groups were as given.
+func (a *App) IfGroupsThenTeamUsersRemoved(teamID string, groupIDs []string, page, perPage int) ([]*model.UserWithGroups, int64, *model.AppError) {
```
Comment (jespino) [inline app/group.go]: Same problem with then naming of the function

File: model/client4.go [hunk 4]
```diff
@@ -4412,3 +4412,15 @@ func (c *Client4) PatchGroupSyncable(groupID, syncableID string, syncableType Gr
 	defer closeBody(r)
 	return GroupSyncableFromJson(r.Body), BuildResponse(r)
 }
+
+func (c *Client4) IfGroupsThenTeamUsersRemoved(teamID string, groupIDs []string, page, perPage int, etag string) ([]*UserWithGroups, int64, *Response) {
```
Comment (jespino) [inline model/client4.go]: Same problem with then naming of the function

File: store/sqlstore/group_store.go [hunk 5]
```diff
@@ -1035,3 +1045,81 @@ func (s *SqlGroupStore) GetGroups(page, perPage int, opts model.GroupSearchOpts)
 
 	return groups, nil
 }
+
+func (s *SqlGroupStore) ifGroupsThenTeamUsersRemovedQuery(teamID string, groupIDs []string, isCount bool) squirrel.SelectBuilder {
```
Comment (jespino) [inline store/sqlstore/group_store.go]: Same problem with then naming of the function

File: store/sqlstore/group_store.go [hunk 6]
```diff
@@ -132,6 +133,15 @@ func (s *SqlGroupStore) Get(groupId string) store.StoreChannel {
 	})
 }
 
+func (s *SqlGroupStore) GetByIDs(groupIDs []string) ([]*model.Group, *model.AppError) {
+	var groups []*model.Group
+	query := fmt.Sprintf("SELECT * from UserGroups WHERE Id IN ('%s')", strings.Join(groupIDs, "', '"))
```
Comment (jespino) [inline store/sqlstore/group_store.go]: Can we use here the squirrel query builder. It solves pretty well the "IN" clauses with lists, and I'm not sure if this change can derive in a sql injection problem, you are concatenating arbitrary strings with ",", and building it into the query with interpolation. Sounds like dangerous to me.

File: store/sqlstore/group_store.go [hunk 7]
```diff
@@ -1035,3 +1045,81 @@ func (s *SqlGroupStore) GetGroups(page, perPage int, opts model.GroupSearchOpts)
...
+	} else {
+		tmpl := "Users.*, TeamMembers.SchemeGuest, TeamMembers.SchemeAdmin, TeamMembers.SchemeUser, %s AS GroupIDs"
+		if s.DriverName() == model.DATABASE_DRIVER_MYSQL {
+			selectStr = fmt.Sprintf(tmpl, "group_concat(UserGroups.Id)")
+		} else {
+			selectStr = fmt.Sprintf(tmpl, "string_agg(UserGroups.Id, ',')")
+		}
+	}
+
+	groupIDsSQL := strings.Join(groupIDs, "', '")
```
Comment (jespino) [inline store/sqlstore/group_store.go]: same problem here with the concatenation and interpolation of this string, I don't trust this kind of stuff, can we try to build it with squirrel?

File: store/sqlstore/group_store.go [hunk 8]
```diff
@@ -1035,3 +1045,81 @@ func (s *SqlGroupStore) GetGroups(page, perPage int, opts model.GroupSearchOpts)
...
+
+	query := s.getQueryBuilder().Select(selectStr).
+		From(`TeamMembers
+			JOIN Teams ON Teams.Id = TeamMembers.TeamId
+			JOIN Users ON Users.Id = TeamMembers.UserId
+			LEFT JOIN Bots ON Bots.UserId = TeamMembers.UserId
+			JOIN GroupMembers ON GroupMembers.UserId = Users.Id
+			JOIN UserGroups ON UserGroups.Id = GroupMembers.GroupId`).
+		Where(fmt.Sprintf(`TeamMembers.DeleteAt = 0
```
Comment (jespino) [inline store/sqlstore/group_store.go]: This is a weird mix where you mix squirrel with Sprintf, why not keep using squirrel here?

File: api4/team.go (hunk 2 above)
Comment (mkraft) [reply api4/team.go]: When you think about it doesn't matter if any hypothetical case is going to happens; that has no bearing on these results now. What this really is is the difference between the **set of team members** and the **set of given group members**, expressed as `Users` records. I'm going to renaming everything with that concept in mind. Perhaps:

`/teams/:team_id/members_minus_group_members`
and
`TeamMembersMinusGroupMembers`

File: store/sqlstore/group_store.go (hunk 6 above)
Comment (mkraft) [reply store/sqlstore/group_store.go]: Good idea.

File: api4/team.go (hunk 2 above)
Comment (jespino) [reply api4/team.go]: Yes, that is very explicit about what it does. I like it.

File: store/sqlstore/group_store.go (hunk 7 above)
Comment (mkraft) [reply store/sqlstore/group_store.go]: I have updated it to be as much in Squirrel as possible but I don't think there's a way to completely eliminate the `fmt.Sprintf` parts without
a) doing something hacky with squirrels args, and
b) doing a conditional on the db driver type

File: store/sqlstore/group_store.go [hunk 9]
```diff
@@ -1035,3 +1050,81 @@ func (s *SqlGroupStore) GetGroups(page, perPage int, opts model.GroupSearchOpts)
...
+		} else {
+			selectStr = fmt.Sprintf(tmpl, "string_agg(UserGroups.Id, ',')")
+		}
+	}
+
+	subQuery := s.getQueryBuilder().Select("GroupMembers.UserId").
+		From("GroupMembers").
+		Join("UserGroups ON UserGroups.Id = GroupMembers.GroupId").
+		Where("GroupMembers.DeleteAt = 0").
+		Where(fmt.Sprintf("GroupMembers.GroupId IN ('%s')", strings.Join(groupIDs, "', '")))
```
Comment (jespino) [inline store/sqlstore/group_store.go]: Still using concatenation, I think we can use here the `sq.Eq` for the comparison here.

File: store/sqlstore/group_store.go [hunk 10]
```diff
@@ -1035,3 +1050,81 @@ func (s *SqlGroupStore) GetGroups(page, perPage int, opts model.GroupSearchOpts)
...
+		Join("Users ON Users.Id = TeamMembers.UserId").
+		LeftJoin("Bots ON Bots.UserId = TeamMembers.UserId").
+		Join("GroupMembers ON GroupMembers.UserId = Users.Id").
+		Join("UserGroups ON UserGroups.Id = GroupMembers.GroupId").
+		Where("TeamMembers.DeleteAt = 0").
+		Where("Teams.DeleteAt = 0").
+		Where("Users.DeleteAt = 0").
+		Where("Bots.UserId IS NULL").
+		Where("Teams.Id = ?", teamID).
+		Where(fmt.Sprintf("Users.Id NOT IN (%s)", sql))
```
Comment (jespino) [inline store/sqlstore/group_store.go]: Same here, I think there should be a way to use squirrel to use subqueries in the where (maybe I'm wrong).

File: store/sqlstore/group_store.go (hunk 9 above)
Comment (mkraft) [reply store/sqlstore/group_store.go]: I guess you didn't see my comment saying that I don't think it's possible.

Comment (jespino) [reply store/sqlstore/group_store.go]: Yes, sorry I didn't see it.

Comment (mkraft) [reply store/sqlstore/group_store.go]: You can create the subquery with argument placeholders ($1, $2...), but then I don't think you can combine the subquery with the main query without manually manipulating the argument and single quotes that it surrounds the subquery with.

Comment (mkraft) [reply store/sqlstore/group_store.go]: There are 183 Squirrel forks. I wonder if any of them have support for subqueries in `WHERE` clauses?

**Pull request merged**